vignettes/symptom-surveys.Rmd

```diff
@@ -0,0 +1,672 @@
+
+# Find "complete" counties, present in all three data signals at all times
+geo_values_complete <- intersect(
+  intersect(g$geo_value, f$geo_value),
+  c$geo_value
+)
+
+# Make one big matrix by joining these three data frames
+z <- full_join(full_join(g, f, by = c("geo_value", "time_value")),
+  c,
+  by = c("geo_value", "time_value")
+) %>%
+  filter(geo_value %in% geo_values_complete) %>%
+  as_epi_df()
+
+Logit <- function(x, a = 0.01) log((x + a) / (1 - x + a))
+Sigmd <- function(y, a = 0.01) (exp(y) * (1 + a) - a) / (1 + exp(y))
+
+#### Parameters #####
+
+# Transforms to consider, in what follows
+trans <- Logit
+inv_trans <- Sigmd
+
+# Rescale factors for our signals: bring them all down to proportions (between
+# 0 and 1)
+rescale_f <- 1e-2 # Originally a percentage
+rescale_g <- 1e-2 # Originally a percentage
+rescale_c <- 1e-5 # Originally a count per 100,000 people
+
+z <- z %>% mutate(
+  case = trans(case * rescale_c),
+  fb = trans(fb * rescale_f),
+  goog = trans(goog * rescale_g)
+)
+
+# lead = 7
+leads <- c(7, 14)
+lags <- c(0, 7, 14)
+n <- 14 # Number of trailing days to use for the training set
+
+# Nine forecast dates
+dates <- seq(as.Date("2020-05-06"), as.Date("2020-05-14"), by = "day")
+
+# List for storage of results
+out_list <- vector(mode = "list", length = length(dates))
+for (k in 1:length(dates)) {
+  date <- dates[k]
+
+  if (date %in% c("2020-05-13", "2020-05-14")) leads <- c(7, 14) else leads <- 7
+
+  # Pre-structuring test data for this date
+  z_te <- z %>%
+    rename(
+      target_date = time_value,
+      target_case = case
+    ) %>%
+    select(geo_value, target_date, target_case)
+
+  # Strawman model
+  out_df0 <- map(leads, ~ flatline_forecaster(
+    z %>%
+      filter(between(time_value, date - .x - max(lags) - n, date)) %>%
+      select(time_value, geo_value, case),
+    outcome = "case",
+    args_list = arx_args_list(
+      lags = lags,
+      ahead = .x,
+      nonneg = FALSE
+    )
+  )$predictions %>%
+    mutate(lead = .x) %>%
+    left_join(z_te %>% filter(target_date == (date + .x)), by = c("geo_value", "target_date"))) %>%
+    list_rbind() %>%
+    mutate(err0 = abs(inv_trans(.pred) - inv_trans(target_case))) %>%
+    select(geo_value, forecast_date, err0, lead)
+
+
+  # Cases model
+  out_df1 <- map(leads, ~ arx_forecaster(
+    z %>%
+      filter(between(time_value, date - .x - max(lags) - n, date)) %>%
+      select(time_value, geo_value, case) %>%
+      filter(complete.cases(.)),
+    outcome = "case",
+    predictors = "case",
+    trainer = quantile_reg(),
+    args_list = arx_args_list(
+      lags = lags,
+      ahead = .x,
+      nonneg = FALSE
+    )
+  )$predictions %>%
+    mutate(lead = .x) %>%
+    left_join(z_te %>% filter(target_date == (date + .x)), by = c("geo_value", "target_date"))) %>%
+    list_rbind() %>%
+    mutate(err1 = abs(inv_trans(.pred) - inv_trans(target_case))) %>%
+    select(geo_value, forecast_date, err1, lead)
```

Comment:
Is it possible to unify this and all the calls below? To avoid rehashing the same code repeatedly?